diff --git a/Makefile b/Makefile
index 62fd0f8..c4dae69 100644
--- a/Makefile
+++ b/Makefile
@@ -139,6 +139,8 @@ UPROGS=\
 	$U/_grind\
 	$U/_wc\
 	$U/_zombie\
+	$U/_trace\
+	$U/_load\
 
 fs.img: mkfs/mkfs README $(UPROGS)
 	mkfs/mkfs fs.img README $(UPROGS)
diff --git a/kernel/param.h b/kernel/param.h
index 6624bff..c6c9cb4 100644
--- a/kernel/param.h
+++ b/kernel/param.h
@@ -11,3 +11,9 @@
 #define NBUF         (MAXOPBLOCKS*3)  // size of disk block cache
 #define FSSIZE       2000  // size of file system in blocks
 #define MAXPATH      128   // maximum file path name
+
+#define MAXSTRARGLEN       128     // keeping same as MAXPATH
+#define MAXSYSCALLARG       6      // since trapframe has a0-a5
+
+#define MAXRECENTCMD    	128 // max prev commands to be stored
+#define MAXCMDLEN           128 // max len of cmd string
diff --git a/kernel/proc.c b/kernel/proc.c
index 58a8a0b..4581464 100644
--- a/kernel/proc.c
+++ b/kernel/proc.c
@@ -26,6 +26,13 @@ extern char trampoline[]; // trampoline.S
 // must be acquired before any p->lock.
 struct spinlock wait_lock;
 
+int total_cmd_count;
+int curr_recent_cmd_index;  // currently at which index... // circular queue
+char recent_cmds[MAXRECENTCMD][MAXCMDLEN];
+// acquire this lock while using the above 3 variables
+struct spinlock cmd_lock;
+
+
 // Allocate a page for each process's kernel stack.
 // Map it high in memory, followed by an invalid
 // guard page.
@@ -48,9 +55,15 @@ void
 procinit(void)
 {
   struct proc *p;
+
+  total_cmd_count = 0;       // the first command not yet arrived
+  curr_recent_cmd_index = 0;
   
   initlock(&pid_lock, "nextpid");
   initlock(&wait_lock, "wait_lock");
+
+  initlock(&cmd_lock, "cmd_lock");
+
   for(p = proc; p < &proc[NPROC]; p++) {
       initlock(&p->lock, "proc");
       p->state = UNUSED;
@@ -146,6 +159,8 @@ found:
   p->context.ra = (uint64)forkret;
   p->context.sp = p->kstack + PGSIZE;
 
+  p->tracenum = -1;
+
   return p;
 }
 
@@ -169,6 +184,8 @@ freeproc(struct proc *p)
   p->killed = 0;
   p->xstate = 0;
   p->state = UNUSED;
+
+  p->tracenum = -1;
 }
 
 // Create a user page table for a given process, with no user memory,
diff --git a/kernel/proc.h b/kernel/proc.h
index d021857..009c23f 100644
--- a/kernel/proc.h
+++ b/kernel/proc.h
@@ -80,6 +80,7 @@ struct trapframe {
 };
 
 enum procstate { UNUSED, USED, SLEEPING, RUNNABLE, RUNNING, ZOMBIE };
+enum syscall_argtype { EMPTY, INT, STR, ADDR };
 
 // Per-process state
 struct proc {
@@ -104,4 +105,22 @@ struct proc {
   struct file *ofile[NOFILE];  // Open files
   struct inode *cwd;           // Current directory
   char name[16];               // Process name (debugging)
+
+  // anik
+
+  int tracenum;                // Number of syscall to be traced
+
+  enum syscall_argtype argtypes[MAXSYSCALLARG];
+
+  int intargs[MAXSYSCALLARG];
+  char strargs[MAXSYSCALLARG][MAXSTRARGLEN];
+  uint64 addrargs[MAXSYSCALLARG];
+
+};
+
+struct procInfo {
+  int activeProcess;  // # of processes in RUNNABLE and RUNNING state
+  int totalProcess;   // # of total possible processes
+  int memsize;        // in bytes; summation of all active process
+  int totalMemSize;   // in bytes; all available physical Memory
 };
diff --git a/kernel/syscall.c b/kernel/syscall.c
index ed65409..96143f6 100644
--- a/kernel/syscall.c
+++ b/kernel/syscall.c
@@ -57,6 +57,14 @@ void
 argint(int n, int *ip)
 {
   *ip = argraw(n);
+
+  struct proc *p = myproc();
+  int num = p->trapframe->a7;
+  if(num == p->tracenum){
+    p->argtypes[n] = INT;
+    p->intargs[n] = *ip;
+  }
+  
 }
 
 // Retrieve an argument as a pointer.
@@ -66,6 +74,14 @@ void
 argaddr(int n, uint64 *ip)
 {
   *ip = argraw(n);
+
+  struct proc *p = myproc();
+  int num = p->trapframe->a7;
+  if(num == p->tracenum){
+    p->argtypes[n] = ADDR;
+    p->addrargs[n] = *ip;
+  }
+
 }
 
 // Fetch the nth word-sized system call argument as a null-terminated string.
@@ -76,7 +92,17 @@ argstr(int n, char *buf, int max)
 {
   uint64 addr;
   argaddr(n, &addr);
-  return fetchstr(addr, buf, max);
+  int ret = fetchstr(addr, buf, max);
+
+  struct proc *p = myproc();
+  int num = p->trapframe->a7;
+  if(num == p->tracenum){
+    p->argtypes[n] = STR;
+    safestrcpy(p->strargs[n], buf, max);
+  }
+
+  return ret;
+
 }
 
 // Prototypes for the functions that handle system calls.
@@ -101,6 +127,37 @@ extern uint64 sys_unlink(void);
 extern uint64 sys_link(void);
 extern uint64 sys_mkdir(void);
 extern uint64 sys_close(void);
+extern uint64 sys_trace(void);
+extern uint64 sys_info(void);
+
+char* syscall_names[] = {
+    [SYS_fork]    "fork",
+    [SYS_exit]    "exit",
+    [SYS_wait]    "wait",
+    [SYS_pipe]    "pipe",
+    [SYS_read]    "read",
+    [SYS_kill]    "kill",
+    [SYS_exec]    "exec",
+    [SYS_fstat]   "fstat",
+    [SYS_chdir]   "chdir",
+    [SYS_dup]     "dup",
+    [SYS_getpid]  "getpid",
+    [SYS_sbrk]    "sbrk",
+    [SYS_sleep]   "sleep",
+    [SYS_uptime]  "uptime",
+    [SYS_open]    "open",
+    [SYS_write]   "write",
+    [SYS_mknod]   "mknod",
+    [SYS_unlink]  "unlink",
+    [SYS_link]    "link",
+    [SYS_mkdir]   "mkdir",
+    [SYS_close]   "close",
+    [SYS_trace]   "trace",
+    [SYS_info]    "info",
+    [SYS_setlastcmd]    "setlastcmd",
+    [SYS_getrecentcmd]  "getrecentcmd",
+};
+
 
 // An array mapping syscall numbers from syscall.h
 // to the function that handles the system call.
@@ -126,6 +183,10 @@ static uint64 (*syscalls[])(void) = {
 [SYS_link]    sys_link,
 [SYS_mkdir]   sys_mkdir,
 [SYS_close]   sys_close,
+[SYS_trace]   sys_trace,
+[SYS_info]    sys_info,
+[SYS_setlastcmd]    sys_setlastcmd,
+[SYS_getrecentcmd]  sys_getrecentcmd,
 };
 
 void
@@ -138,7 +199,57 @@ syscall(void)
   if(num > 0 && num < NELEM(syscalls) && syscalls[num]) {
     // Use num to lookup the system call function for num, call it,
     // and store its return value in p->trapframe->a0
+
+    int istraced = num == p->tracenum;
+
+    int i;
+    for(i=0; i<MAXSYSCALLARG; i++){
+      p->argtypes[i] = EMPTY;   // at the beginning of each syscall, first empty the argtypes...
+    }
+
     p->trapframe->a0 = syscalls[num]();
+
+    if(istraced){
+      acquire(&p->lock);
+      int pid = p->pid;
+      release(&p->lock);
+
+      char*syscall_name = syscall_names[num];
+
+      int return_value = p->trapframe->a0;
+
+      printf("pid: %d, syscall: %s, args:(", pid, syscall_name);
+
+      int args_ended = 0;
+
+      for(i=0; i<MAXSYSCALLARG; i++){
+        switch (p->argtypes[i])
+        {
+          case INT:
+            printf("%d", p->intargs[i]);
+            break;
+          case ADDR:
+            printf("%p", (void *)p->addrargs[i]);
+            break;
+          case STR:
+            printf("%s", p->strargs[i]);
+            break;
+          default:
+            args_ended = 1;
+            break;
+        }
+        if(args_ended == 1){
+          break;
+        } else {
+          if((i < MAXSYSCALLARG-1) && (p->argtypes[i+1] != EMPTY)){
+            printf(", ");
+          }
+        }
+      }
+
+      printf("), return: %d\n", return_value);
+    }
+
   } else {
     printf("%d %s: unknown sys call %d\n",
             p->pid, p->name, num);
diff --git a/kernel/syscall.h b/kernel/syscall.h
index bc5f356..505766b 100644
--- a/kernel/syscall.h
+++ b/kernel/syscall.h
@@ -20,3 +20,7 @@
 #define SYS_link   19
 #define SYS_mkdir  20
 #define SYS_close  21
+#define SYS_trace  22
+#define SYS_info   23
+#define SYS_setlastcmd   24
+#define SYS_getrecentcmd   25
diff --git a/kernel/sysproc.c b/kernel/sysproc.c
index 3b4d5bd..e552df9 100644
--- a/kernel/sysproc.c
+++ b/kernel/sysproc.c
@@ -91,3 +91,107 @@ sys_uptime(void)
   release(&tickslock);
   return xticks;
 }
+
+// traces a specific system call over all processes
+uint64
+sys_trace(void)
+{
+  int num;
+  argint(0, &num);
+
+  struct proc* curr_proc = myproc();
+
+  acquire(&curr_proc->lock);
+  curr_proc->tracenum = num;  // set the property in the struct of proc
+  release(&curr_proc->lock);
+
+  // signal that tracing is successfully enabled
+  return 0;
+}
+
+
+uint64
+sys_info(void)
+{
+  uint64 addr;
+  argaddr(0, &addr);
+
+  struct procInfo pinfo = {
+    .activeProcess = 0,
+    .totalProcess = 0,
+    .memsize = 0,
+    .totalMemSize = 0
+  };
+
+  struct proc *p;
+  extern struct proc proc[NPROC];
+  for(p = proc; p < &proc[NPROC]; p++){
+    acquire(&p->lock);
+    enum procstate pstate = p->state;
+    release(&p->lock);
+    if(pstate == RUNNABLE || pstate == RUNNING){
+      pinfo.memsize += p->sz;
+      pinfo.activeProcess++;
+    }
+  }
+
+  pinfo.totalProcess = NPROC;
+  pinfo.totalMemSize = PHYSTOP - KERNBASE;
+
+  if(copyout(myproc()->pagetable, addr, (char *)&pinfo, sizeof(pinfo)) < 0){
+    return -1;
+  }
+
+  return 0;
+}
+
+uint64
+sys_setlastcmd(void)
+{
+  char cmd[MAXCMDLEN];
+
+  if(argstr(0, cmd, MAXCMDLEN) < 0)
+    return -1;
+
+  extern int total_cmd_count;
+  extern int curr_recent_cmd_index;
+  extern char recent_cmds[MAXRECENTCMD][MAXCMDLEN];
+  extern struct spinlock cmd_lock;
+
+  acquire(&cmd_lock);
+  total_cmd_count++;
+  safestrcpy(recent_cmds[curr_recent_cmd_index], cmd, MAXCMDLEN);
+  curr_recent_cmd_index = (curr_recent_cmd_index + 1)%MAXRECENTCMD; // circular queue
+  release(&cmd_lock);
+
+  return 0;
+}
+
+
+uint64
+sys_getrecentcmd(void)
+{
+  int n;
+
+  argint(0,&n);
+
+  char cmd[MAXCMDLEN];
+
+  extern int total_cmd_count;
+  extern int curr_recent_cmd_index;
+  extern char recent_cmds[MAXRECENTCMD][MAXCMDLEN];
+  extern struct spinlock cmd_lock;
+
+  if(n > total_cmd_count){
+    return -1;
+  }
+
+  int target_index = (curr_recent_cmd_index - n + 1 + MAXRECENTCMD) % MAXRECENTCMD;
+
+  acquire(&cmd_lock);
+  safestrcpy(cmd, recent_cmds[target_index], cmd, MAXCMDLEN);
+  printf("wanted command = %s\n", cmd);
+  release(&cmd_lock);
+
+  return 0;
+}
diff --git a/user/sh.c b/user/sh.c
index 836ebcb..b3f14a0 100644
--- a/user/sh.c
+++ b/user/sh.c
@@ -158,6 +158,12 @@ main(void)
 
   // Read and run input commands.
   while(getcmd(buf, sizeof(buf)) >= 0){
+    if(buf[0] == '!' && buf[1] == '!' && buf[2] == ' '){
+      buf[strlen(buf)-1] = 0;  // chop \n
+      int n = atoi(buf+3);
+      getrecentcmd(n);
+      continue;
+    }
     if(buf[0] == 'c' && buf[1] == 'd' && buf[2] == ' '){
       // Chdir must be called by the parent, not the child.
       buf[strlen(buf)-1] = 0;  // chop \n
@@ -167,6 +173,7 @@ main(void)
     }
     if(fork1() == 0)
       runcmd(parsecmd(buf));
+    setlastcmd(buf);
     wait(0);
   }
   exit(0);
diff --git a/user/trace.c b/user/trace.c
new file mode 100644
index 0000000..f59468a
--- /dev/null
+++ b/user/trace.c
@@ -0,0 +1,29 @@
+#include "kernel/param.h"
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int main(int argc, char *argv[])
+{
+    int i;
+    char *nargv[MAXARG];
+
+    if (argc < 3 || (argv[1][0] < '0' || argv[1][0] > '9'))
+    {
+        fprintf(2, "Usage: %s sys_call_num command\n", argv[0]);
+        exit(1);
+    }
+
+    if (trace(atoi(argv[1])) < 0)
+    {
+        fprintf(2, "%s: trace failed\n", argv[0]);
+        exit(1);
+    }
+
+    for (i = 2; i < argc && i < MAXARG; i++)
+    {
+        nargv[i - 2] = argv[i];
+    }
+    exec(nargv[0], nargv);
+    exit(0);
+}
\ No newline at end of file
diff --git a/user/user.h b/user/user.h
index 04013ca..46c371c 100644
--- a/user/user.h
+++ b/user/user.h
@@ -1,5 +1,12 @@
 struct stat;
 
+struct procInfo {
+    int activeProcess;  // # of processes in RUNNABLE and RUNNING state
+    int totalProcess;   // # of total possible processes
+    int memsize;        // in bytes; summation of all active process
+    int totalMemSize;   // in bytes; all available physical Memory
+};
+
 // system calls
 int fork(void);
 int exit(int) __attribute__((noreturn));
@@ -22,6 +29,10 @@ int getpid(void);
 char* sbrk(int);
 int sleep(int);
 int uptime(void);
+int trace(int);
+int info(struct procInfo*);
+int setlastcmd(const char*);
+int getrecentcmd(int);
 
 // ulib.c
 int stat(const char*, struct stat*);
diff --git a/user/usys.pl b/user/usys.pl
index 01e426e..1344d0f 100755
--- a/user/usys.pl
+++ b/user/usys.pl
@@ -36,3 +36,7 @@ entry("getpid");
 entry("sbrk");
 entry("sleep");
 entry("uptime");
+entry("trace");
+entry("info");
+entry("setlastcmd");
+entry("getrecentcmd");
